### PR TITLE
fix for SI-5672

### PR DESCRIPTION
**Pull request by magarciaEPFL**
review by @VladUreche

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/icode/Members.scala [hunk 1]
```diff
@@ -257,11 +257,23 @@ trait Members {
...
-            code.removeBlock(succ)
+            val lastInstr = bb.lastInstruction
+            /* Ticket SI-5672
+             * Besides removing the control-flow instruction at the end of `bb` (usually a JUMP), we have to pop any values it pushes.
+             * Examples:
+             *   `SWITCH` consisting of just the default case, or
+             *   `CJUMP(targetBlock, targetBlock, _, _)` ie where success and failure targets coincide (this one consumes two stack values).
+             */
+            val oldTKs = lastInstr.consumedTypes
+            assert(lastInstr.consumed == oldTKs.size, "Someone forgot to override consumedTypes() in " +  lastInstr)
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/backend/icode/Members.scala:268]: This assertion should probably go inside the constructor of `Instruction`.

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 2]
```diff
@@ -1646,7 +1646,7 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
       // TODO: if patterns allow switch but the type of the scrutinee doesn't, cast (type-test) the scrutinee to the corresponding switchable type and switch on the result
       if (regularSwitchMaker.switchableTpe(scrutSym.tpe)) {
         val caseDefsWithDefault = regularSwitchMaker(cases map {c => (scrutSym, c)}, pt)
-        if (caseDefsWithDefault.length <= 2) None // not worth emitting a switch... also, the optimizer has trouble digesting tiny switches, apparently, so let's be nice and not generate them
+        if (caseDefsWithDefault isEmpty) None // not worth emitting a switch.
```
Comment (dragos) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:1649]: So is the optimizer now happy with one-branch switches?

Comment (magarciaEPFL) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:1649]: Yes, the changes above result in that one-branch-switch becoming straight-line code, [a comment in SI-5672](https://issues.scala-lang.org/browse/SI-5672?focusedCommentId=57812&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-57812) shows the javap output for it.

File: src/compiler/scala/tools/nsc/backend/icode/Members.scala (hunk 1 above)
Comment (magarciaEPFL) [inline src/compiler/scala/tools/nsc/backend/icode/Members.scala:268]: Yes, that's the right thing to do but we aren't there yet. I've created a ticket for that: https://issues.scala-lang.org/browse/SI-5819

**Comment by VladUreche** [comment]
LGTM, go ahead with the merge :))

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala (hunk 2 above)
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala:1649]: great!

**Pull request merged**